feat: API endpoint to export usage data into CSV [DEV-4754]

filipdjokic opened the pull request.

filipdjokic committed:
feat: API endpoint to export usage data into CSV

ankurdotb commented:
Task linked: [DEV-4754 Create /export endpoint to fetch usage data](https://app.clickup.com/t/6600954/DEV-4754)

ankurdotb commented on packages/faucet/src/api/webserver.ts:
```diff
@@ -97,6 +97,64 @@ export class Webserver {
+            throw new HttpError(405, "This endpoint requires a GET request");
+          }
+
+          // Parse query parameters to sort by start and end date
```
1. Ideally should just be able to send dates as YYYY-MM-DD with no HH:MM. If sent as fully qualified with HH:MM or timezone data, in any case, normalise UTC / Zulu time.
2. Validate startDate and endDate are actually dates, endDate is later than startDate, and startDate is older than current time in UTC (all failure modes)
3. If it fails validation, reject early with helpful error messages (not sure what current catch/throw error will do).

filipdjokic replied:
All the comments were addressed and now API:

1. Accepts both YYYY-MM-DD and Zulu / ISO 8601
2. Performing additional validation including all examples stated above
3. The error messages are more user-friendly.

ankurdotb reviewed (changes requested):
Also as stated in the ClickUp task, if there's no `startDate` and `endDate` defined, export the last 30 days (rather than all).

filipdjokic committed:
Include additional normalization and validation

filipdjokic committed:
Cleanup

ankurdotb committed:
Cursor-driven error handling

Better Query Validation:
Added type safety with ExportRequestQuery interface
Added explicit query validation before processing
Proper type checking for query parameters
Enhanced Error Handling:
Using the expose parameter of HttpError to control error message visibility
More specific error handling for date validation vs other errors
Better error logging for debugging
Security Improvements:
Added no-cache headers to prevent sensitive data caching
Better CSV field sanitization to prevent CSV injection
Consistent error messages that don't leak internal details
Better Date Range Handling:
More descriptive filename based on date range
Leveraging existing validateDateRange function from utils
No need for separate date validation as it's handled by validateDateRange
Code Organization:
Separated query validation into its own method
More consistent error handling pattern
Better type safety throughout

ankurdotb committed:
Cursor-driven date handling improvements

Better date normalization:
Accepts both YYYY-MM-DD and full ISO formats
Converts YYYY-MM-DD to UTC midnight of that day
Clear error messages for invalid formats
Clearer date range validation:
Default 30-day range when no dates provided
Handles all cases (no dates, start only, end only, both dates)
Prevents future dates
Ensures start date is before end date
Better error messages:
More specific about what went wrong
Includes current UTC time in future date errors
Shows actual date strings in error messages
4. Type safety:
Clear return types
Proper error handling
Consistent date object usage

filipdjokic committed:
Modify filename logic

ankurdotb committed:
run format
